jwt/jwt.go

```diff
@@ -2,55 +2,121 @@ package jwt
-// context. If we don't use a custom unexported type it's easy for others
-// to get our value or accidentally overwrite it.
-type ctxValue string
+type (
+	// ctxValue is a helper type which makes it safe to register values in the
+	// context. If we don't use a custom unexported type it's easy for others
+	// to get our value or accidentally overwrite it.
+	ctxValue string
+
+	// tokenSession is the bare minimum we need in the `session` claim in our
+	// JWTs.
+	tokenSession struct {
+		Active   bool          `json:"active"`
+		Identity tokenIdentity `json:"identity"`
+	}
+	tokenIdentity struct {
+		Traits tokenTraits `json:"traits"`
+	}
+	tokenTraits struct {
+		Email string `json:"email"`
+	}
+)
+
+// AccountsPublicKeySet exposes the public version of the JWKS we use for
+// signing JWTs. This allows anyone to fetch the public key and verify that the
+// tokens they receive are issued by us.
+func AccountsPublicKeySet(logger *logrus.Logger) (jwk.Set, error) {
+	ks, err := accountsKeySet(logger)
+	if err != nil {
+		return nil, err
+	}
+	return jwk.PublicSetOf(ks)
+}
 
 // ContextWithToken returns a copy of the given context that contains a token.
 func ContextWithToken(ctx context.Context, token *jwt.Token) context.Context {
 	return context.WithValue(ctx, ctxValue("token"), token)
 }
 
-// TokenExpiration extracts and returns the `exp` claim of the given token.
-// NOTE: It does NOT validate the token!
-func TokenExpiration(t *jwt.Token) (int64, error) {
-	if t == nil {
-		return 0, errors.New("invalid token")
+// TokenForUser creates a serialized JWT token for the given user.
+//
+// The tokens generated by this function are a slimmed down version of the ones
+// described in ValidateToken's docstring.
+func TokenForUser(logger *logrus.Logger, email, sub string) (jwt.Token, []byte, error) {
+	keySet, err := accountsKeySet(logger)
+	if err != nil {
+		return nil, nil, errors.AddContext(err, "failed to fetch the JWKS")
 	}
-	if reflect.ValueOf(t.Claims).Kind() != reflect.ValueOf(jwt.MapClaims{}).Kind() {
-		return 0, errors.New("the token does not contain the claims we expect")
+	key, found := keySet.Get(0)
+	if !found {
+		return nil, nil, errors.New("JWKS is empty")
 	}
-	claims := t.Claims.(jwt.MapClaims)
-	if reflect.ValueOf(claims["exp"]).Kind() != reflect.Float64 {
-		return 0, errors.New("the token does not contain the claims we expect")
+	t, err := tokenForUser(email, sub)
+	if err != nil {
+		return nil, nil, errors.AddContext(err, "failed to build token")
+	}
+	var sigAlgo jwa.SignatureAlgorithm
+	for _, sa := range jwa.SignatureAlgorithms() {
+		if string(sa) == key.Algorithm() {
+			sigAlgo = sa
+			break
+		}
 	}
-	return int64(claims["exp"].(float64)), nil
+	if sigAlgo == "" {
+		return nil, nil, errors.New("failed to determine signature algorithm")
+	}
+	bytes, err := jwt.Sign(t, sigAlgo, key)
+	if err != nil {
+		return nil, nil, errors.New("failed to sign token")
+	}
+	tk, err := jwt.Parse(bytes)
```

Comment:
Seems like you never need the parsed result so I wouldn't bother parsing and returning it

Reply:
Oh, this is an artifact of splitting the code into several PRs. This function is not actually used in this PR and should have been in the next one where you can find this line where we use the token to grab its expiration time:
https://github.com/SkynetLabs/skynet-accounts/pull/62/files#diff-855b4003387cd9e2b8e2c353c747fd1817eccf1d223b4066f0d7a1db0b4b4b4eR109